modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/RuntimeMessageHandlerImpl.java

```diff
@@ -86,56 +89,37 @@ public void beforeDeactivation(ComponentManager mgr) {
     }
 
     protected void changeStep(ComponentManagerStep step) {
-        messages.removeIf(msg -> step.equals(msg.getStep()));
-        this.step = step;
+        this.currentStep = step;
     }
 
-    /**
-     * @since 9.10
-     */
-    protected static class Message {
-
-        protected final ComponentManagerStep step;
-
-        protected final Level level;
-
-        protected final String message;
-
-        public Message(ComponentManagerStep step, Level level, String message) {
-            this.step = step;
-            this.level = level;
-            this.message = message;
-        }
-
-        public ComponentManagerStep getStep() {
-            return step;
-        }
-
-        public Level getLevel() {
-            return level;
-        }
-
-        public String getMessage() {
-            return message;
-        }
-
+    @Override
+    public void addMessage(RuntimeMessage message) {
+        messages.add(message);
     }
 
-    /**
-     * @since 9.10
-     */
-    protected enum ComponentManagerStep {
-
-        ACTIVATING,
-
-        STARTING,
-
-        RUNNING,
+    @Override
+    public void addMessage(Level level, String message, Class<?> source) {
+        addMessage(new RuntimeMessage(currentStep, level, message, source == null ? null : source.getName()));
+    }
 
-        STOPPING,
+    @Override
+    public List<RuntimeMessage> getMessages(Predicate<RuntimeMessage> predicate) {
+        return messages.stream()
+                       .filter(msg -> predicate.test(msg))
+                       .collect(Collectors.collectingAndThen(Collectors.toList(), Collections::unmodifiableList));
```

Comment:
I know it's a refactoring of pre-existing code, but FYI since Java 10 you can directly do `.collect(Collectors.toUnmodifiableList())`

Reply:
:+1: